File: src/jqmas/parser/parser.py
"""Parser de Jqmas — convierte tokens en AST."""

from jqmas.lexer.token import Token, TokenType
from jqmas.parser.ast_nodes import (
    Argument,
    BinaryOp,
    NumberLiteral,
    PresCall,
    Program,
    StringLiteral,
)


class ParserError(SyntaxError):
    """Error durante el análisis sintáctico."""


class Parser:
    """Parser de descenso recursivo para Jqmas."""

    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.pos = 0

    # ----- helpers -----

    def _peek(self) -> Token:
        return self.tokens[self.pos]

    def _previous(self) -> Token:
        return self.tokens[self.pos - 1]

    def _advance(self) -> Token:
        if not self._at_end():
            self.pos += 1
        return self._previous()

    def _at_end(self) -> bool:
        return self._peek().type == TokenType.EOF

    def _check(self, type_: TokenType) -> bool:
        return self._peek().type == type_

    def _match(self, *types: TokenType) -> bool:
        for t in types:
            if self._check(t):
                self._advance()
                return True
        return False

    def _consume(self, type_: TokenType, msg: str) -> Token:
        if self._check(type_):
            return self._advance()
        tok = self._peek()
        raise ParserError(
            f"{msg} (encontrado {tok.type.name} en línea {tok.line}, columna {tok.column})"
        )

    def _error(self, msg: str) -> None:
        tok = self._peek()
        raise ParserError(f"{msg} (línea {tok.line}, columna {tok.column})")

    # ----- parse -----

    def parse(self) -> Program:
        """Parsea el programa completo."""
        statements = []
        while not self._at_end():
            statements.append(self._statement())
        return Program(statements=statements)

    def _statement(self) -> object:
        """Una sentencia: por ahora solo pres[...]."""
        if self._check(TokenType.PRES):
            return self._pres_call()
        self._error("Se esperaba 'pres'")

    def _pres_call(self) -> PresCall:
        """Parsea pres[args...]."""
        self._consume(TokenType.PRES, "Se esperaba 'pres'")
        self._consume(TokenType.LBRACKET, "Se esperaba '[' después de pres")

        arguments: list[Argument] = []

        # Caso especial: pres[] vacío
        if not self._check(TokenType.RBRACKET):
            arguments.append(self._argument())
            while self._match(TokenType.COMMA):
                arguments.append(self._argument())

        self._consume(TokenType.RBRACKET, "Se esperaba ']' para cerrar pres")
        return PresCall(arguments=arguments)

    def _argument(self) -> Argument:
        """Parsea un argumento, con posible prefijo notN/."""
        not_count = 0

        # Detectar notN/ (not, opcionalmente número, slash)
        if self._check(TokenType.NOT):
            self._advance()  # consume 'not'

            # Número opcional entre not y /
            if self._check(TokenType.NUMBER):
                num_tok = self._advance()
                n = num_tok.value
                if not isinstance(n, int):
                    raise ParserError(
                        f"notN/ solo acepta número entero, no decimal (línea {num_tok.line})"
                    )
                if n < 1:
                    raise ParserError(
                        f"not0/ no es válido: el mínimo es 1 (línea {num_tok.line})"
                    )
                if n > 34:
                    raise ParserError(
                        f"not{n}/ excede el límite de 34 saltos (línea {num_tok.line})"
                    )
                not_count = n
            else:
                not_count = 1

            # El slash obligatorio
            self._consume(TokenType.SLASH, "Se esperaba '/' después de not")

            # Si lo que sigue es ] o , entonces notN/ va sin argumento: no hace nada
            if self._check(TokenType.RBRACKET) or self._check(TokenType.COMMA):
                return Argument(value=NumberLiteral(0), not_count=0)

        # Parsear la expresión del argumento
        value = self._expression()
        return Argument(value=value, not_count=not_count)

    # ----- expresiones -----

    def _expression(self) -> object:
        return self._addition()

    def _addition(self) -> object:
        left = self._multiplication()
        while self._match(TokenType.PLUS, TokenType.MINUS):
            op_tok = self._previous()
            op = "+" if op_tok.type == TokenType.PLUS else "-"
            right = self._multiplication()
            left = BinaryOp(left=left, operator=op, right=right)
        return left

    def _multiplication(self) -> object:
        left = self._unary()
        while self._match(TokenType.STAR, TokenType.SLASH):
            op_tok = self._previous()
            op = "*" if op_tok.type == TokenType.STAR else "/"
            right = self._unary()
            left = BinaryOp(left=left, operator=op, right=right)
        return left

    def _unary(self) -> object:
        # Acepta -expr como número negativo (útil)
        if self._match(TokenType.MINUS):
            operand = self._unary()
            return BinaryOp(
                left=NumberLiteral(0),
                operator="-",
                right=operand,
            )
        return self._primary()

    def _primary(self) -> object:
        if self._match(TokenType.NUMBER):
            return NumberLiteral(self._previous().value)
        if self._match(TokenType.STRING):
            return StringLiteral(self._previous().value)
        if self._match(TokenType.LPAREN):
            expr = self._expression()
            self._consume(TokenType.RPAREN, "Se esperaba ')'")
            return expr
        self._error("Se esperaba una expresión")

File: go/internal/parser/parser.go
package parser

import (
"fmt"

"github.com/lucassogarayDA/jqmas/go/internal/lexer"
)

// ParserError es un error de sintaxis con info para formatear.
type ParserError struct {
Mensaje     string
Linea       int
Columna     int
Sugerencias []string
Tips        []string
}

func (e *ParserError) Error() string {
return e.Mensaje
}

var keywordsValidas = []string{"pres", "haz", "not"}

// Parser convierte tokens en un AST.
type Parser struct {
tokens []lexer.Token
pos    int
}

// NewParser crea un parser con la lista de tokens.
func NewParser(tokens []lexer.Token) *Parser {
return &Parser{tokens: tokens, pos: 0}
}

func (p *Parser) peek() lexer.Token {
return p.tokens[p.pos]
}

func (p *Parser) previous() lexer.Token {
return p.tokens[p.pos-1]
}

func (p *Parser) advance() lexer.Token {
if !p.atEnd() {
p.pos++
}
return p.previous()
}

func (p *Parser) atEnd() bool {
return p.peek().Type == lexer.EOF
}

func (p *Parser) check(t lexer.TokenType) bool {
return p.peek().Type == t
}

func (p *Parser) match(types ...lexer.TokenType) bool {
for _, t := range types {
if p.check(t) {
p.advance()
return true
}
}
return false
}

func (p *Parser) consume(t lexer.TokenType, msg string, tips []string) (lexer.Token, error) {
if p.check(t) {
return p.advance(), nil
}
tok := p.peek()
sugerenciasTips := []string{}
sugerenciasTips = append(sugerenciasTips, tips...)

switch t {
case lexer.RBRACKET:
sugerenciasTips = append(sugerenciasTips, "revisá que cada '[' tenga su ']'")
case lexer.RPAREN:
sugerenciasTips = append(sugerenciasTips, "revisá que cada '(' tenga su ')'")
}

return lexer.Token{}, &ParserError{
Mensaje: msg,
Linea:   tok.Line,
Columna: tok.Col,
Tips:    sugerenciasTips,
}
}

// Parse convierte los tokens en un Program.
func (p *Parser) Parse() (*Program, error) {
statements := []Node{}
for !p.atEnd() {
stmt, err := p.statement()
if err != nil {
return nil, err
}
statements = append(statements, stmt)
}
return &Program{Statements: statements}, nil
}

func (p *Parser) statement() (Node, error) {
if p.check(lexer.PRES) {
return p.presCall()
}
if p.check(lexer.HAZ) {
return p.hazAssignment()
}

tok := p.peek()
if tok.Type == lexer.IDENTIFIER {
name := tok.Value.(string)
sugerencias := buscarSimilar(name, keywordsValidas)
if len(sugerencias) > 0 {
return nil, &ParserError{
Mensaje:     fmt.Sprintf("'%s' no es una instrucción válida", name),
Linea:       tok.Line,
Columna:     tok.Col,
Sugerencias: sugerencias,
Tips:        []string{"las instrucciones válidas son: pres, haz"},
}
}
return nil, &ParserError{
Mensaje: fmt.Sprintf("'%s' no es una instrucción válida", name),
Linea:   tok.Line,
Columna: tok.Col,
Tips:    []string{"las instrucciones válidas son: pres, haz"},
}
}

return nil, &ParserError{
Mensaje: "Se esperaba 'pres' o 'haz'",
Linea:   tok.Line,
Columna: tok.Col,
Tips:    []string{"una instrucción empieza con 'pres[' o con 'haz['"},
}
}

func (p *Parser) hazAssignment() (Node, error) {
if _, err := p.consume(lexer.HAZ, "Se esperaba 'haz'", nil); err != nil {
return nil, err
}
if _, err := p.consume(lexer.LBRACKET, "Se esperaba '[' después de haz",
[]string{"la sintaxis es haz[nombre = valor]"}); err != nil {
return nil, err
}

// Nombre de variable
if !p.check(lexer.IDENTIFIER) {
tok := p.peek()
if tok.Type == lexer.NUMBER {
return nil, &ParserError{
Mensaje: "El nombre de variable no puede ser un número",
Linea:   tok.Line,
Columna: tok.Col,
Tips: []string{
"los nombres empiezan con letra o _",
"por ejemplo: haz[x = 5], haz[contador = 0]",
},
}
}
if tok.Type == lexer.STRING {
return nil, &ParserError{
Mensaje: "El nombre de variable no puede ser un string",
Linea:   tok.Line,
Columna: tok.Col,
Tips: []string{
"los nombres van sin comillas",
"por ejemplo: haz[nombre = 'Lucas']",
},
}
}
return nil, &ParserError{
Mensaje: "Se esperaba un nombre de variable",
Linea:   tok.Line,
Columna: tok.Col,
Tips:    []string{"la sintaxis es haz[nombre = valor]"},
}
}

nameTok := p.advance()
name := nameTok.Value.(string)

// Igual
if !p.check(lexer.EQUALS) {
tok := p.peek()
if tok.Type == lexer.RBRACKET {
return nil, &ParserError{
Mensaje: "Falta el '= valor' en la asignación",
Linea:   tok.Line,
Columna: tok.Col,
Tips: []string{
fmt.Sprintf("escribiste: haz[%s]", name),
fmt.Sprintf("la sintaxis correcta es: haz[%s = valor]", name),
},
}
}
return nil, &ParserError{
Mensaje: "Se esperaba '=' después del nombre de variable",
Linea:   tok.Line,
Columna: tok.Col,
Tips:    []string{"la sintaxis es haz[nombre = valor]"},
}
}

p.advance() // consume =

value, err := p.expression()
if err != nil {
return nil, err
}

if _, err := p.consume(lexer.RBRACKET, "Se esperaba ']' para cerrar haz",
[]string{"la sintaxis es haz[nombre = valor]"}); err != nil {
return nil, err
}

return &Assignment{Name: name, Value: value}, nil
}

func (p *Parser) presCall() (Node, error) {
if _, err := p.consume(lexer.PRES, "Se esperaba 'pres'", nil); err != nil {
return nil, err
}
if _, err := p.consume(lexer.LBRACKET, "Se esperaba '[' después de pres",
[]string{"la sintaxis es pres[expresión]"}); err != nil {
return nil, err
}

args := []Argument{}
if !p.check(lexer.RBRACKET) {
arg, err := p.argument()
if err != nil {
return nil, err
}
args = append(args, arg)
for p.match(lexer.COMMA) {
arg, err := p.argument()
if err != nil {
return nil, err
}
args = append(args, arg)
}
}

if _, err := p.consume(lexer.RBRACKET, "Se esperaba ']' para cerrar pres",
[]string{"la sintaxis es pres[expresión]"}); err != nil {
return nil, err
}

return &PresCall{Arguments: args}, nil
}

func (p *Parser) argument() (Argument, error) {
notCount := 0

if p.check(lexer.NOT) {
p.advance()

if p.check(lexer.NUMBER) {
numTok := p.advance()
n := numTok.Value.(int)
if n < 1 {
return Argument{}, &ParserError{
Mensaje: "not0/ no es válido: el mínimo es 1",
Linea:   numTok.Line,
Columna: numTok.Col,
Tips: []string{
"para 1 salto usá not/ o not1/",
"para 2 saltos usá not2/",
},
}
}
if n > 34 {
return Argument{}, &ParserError{
Mensaje: fmt.Sprintf("not%d/ excede el límite de 34 saltos", n),
Linea:   numTok.Line,
Columna: numTok.Col,
Tips:    []string{"el máximo es not34/"},
}
}
notCount = n
} else {
notCount = 1
}

if _, err := p.consume(lexer.SLASH, "Se esperaba '/' después de not",
[]string{"la sintaxis es notN/ antes de un argumento"}); err != nil {
return Argument{}, err
}

if p.check(lexer.RBRACKET) || p.check(lexer.COMMA) {
return Argument{Value: &NumberLiteral{Value: 0}, NotCount: 0}, nil
}
}

value, err := p.expression()
if err != nil {
return Argument{}, err
}
return Argument{Value: value, NotCount: notCount}, nil
}

func (p *Parser) expression() (Node, error) {
return p.addition()
}

func (p *Parser) addition() (Node, error) {
left, err := p.multiplication()
if err != nil {
return nil, err
}
for p.match(lexer.PLUS, lexer.MINUS) {
opTok := p.previous()
op := "+"
if opTok.Type == lexer.MINUS {
op = "-"
}
right, err := p.multiplication()
if err != nil {
return nil, err
}
left = &BinaryOp{Left: left, Operator: op, Right: right}
}
return left, nil
}

func (p *Parser) multiplication() (Node, error) {
left, err := p.unary()
if err != nil {
return nil, err
}
for p.match(lexer.STAR, lexer.SLASH) {
opTok := p.previous()
op := "*"
if opTok.Type == lexer.SLASH {
op = "/"
}
right, err := p.unary()
if err != nil {
return nil, err
}
left = &BinaryOp{Left: left, Operator: op, Right: right}
}
return left, nil
}

func (p *Parser) unary() (Node, error) {
if p.match(lexer.MINUS) {
operand, err := p.unary()
if err != nil {
return nil, err
}
return &BinaryOp{Left: &NumberLiteral{Value: 0}, Operator: "-", Right: operand}, nil
}
return p.primary()
}

func (p *Parser) primary() (Node, error) {
if p.match(lexer.NUMBER) {
return &NumberLiteral{Value: toFloat(p.previous().Value)}, nil
}
if p.match(lexer.STRING) {
return &StringLiteral{Value: p.previous().Value.(string)}, nil
}
if p.match(lexer.IDENTIFIER) {
return &Variable{Name: p.previous().Value.(string)}, nil
}
if p.match(lexer.LPAREN) {
expr, err := p.expression()
if err != nil {
return nil, err
}
if _, err := p.consume(lexer.RPAREN, "Se esperaba ')'", nil); err != nil {
return nil, err
}
return expr, nil
}

tok := p.peek()
tips := []string{"una expresión puede ser: número, string, variable, o a + b"}
if tok.Type == lexer.PRES {
tips = append([]string{"pres[...] no va dentro de una expresión"}, tips...)
} else if tok.Type == lexer.HAZ {
tips = append([]string{"haz[...] es una instrucción, no una expresión"}, tips...)
}

return nil, &ParserError{
Mensaje: "Se esperaba una expresión",
Linea:   tok.Line,
Columna: tok.Col,
Tips:    tips,
}
}

// buscarSimilar busca strings similares a `nombre` en `opciones`.
func buscarSimilar(nombre string, opciones []string) []string {
nombreLower := toLower(nombre)

// Casos específicos
especificas := map[string]string{
"print": "pres", "echo": "pres", "puts": "pres", "printf": "pres",
"pres_": "pres", "pre": "pres", "prs": "pres",
"has": "haz", "haz_": "haz", "hac": "haz",
"nt": "not", "n0t": "not",
}
if sug, ok := especificas[nombreLower]; ok {
for _, op := range opciones {
if op == sug {
return []string{sug}
}
}
}

// Búsqueda genérica simple (Levenshtein simplificado)
resultado := []string{}
for _, op := range opciones {
if distanciaLevenshtein(nombreLower, toLower(op)) <= 2 {
resultado = append(resultado, op)
}
}
return resultado
}

func toLower(s string) string {
result := []byte(s)
for i := range result {
if result[i] >= 'A' && result[i] <= 'Z' {
result[i] += 32
}
}
return string(result)
}

func distanciaLevenshtein(a, b string) int {
la, lb := len(a), len(b)
if la == 0 {
return lb
}
if lb == 0 {
return la
}
prev := make([]int, lb+1)
for j := 0; j <= lb; j++ {
prev[j] = j
}
for i := 1; i <= la; i++ {
curr := make([]int, lb+1)
curr[0] = i
for j := 1; j <= lb; j++ {
costo := 1
if a[i-1] == b[j-1] {
costo = 0
}
curr[j] = min3(curr[j-1]+1, prev[j]+1, prev[j-1]+costo)
}
prev = curr
}
return prev[lb]
}

func min3(a, b, c int) int {
if a < b {
if a < c {
return a
}
return c
}
if b < c {
return b
}
return c
}
func toFloat(v interface{}) float64 {
	switch n := v.(type) {
	case int:
		return float64(n)
	case float64:
		return n
	}
	return 0
}
